cpp/src/strings/replace/replace.cu

```diff
@@ -80,33 +102,437 @@ struct replace_fn {
+  rmm::device_uvector<size_type> target_positions(target_count, stream);
+  auto d_target_positions = target_positions.data();
+  thrust::copy_if(rmm::exec_policy(stream),
+                  thrust::make_counting_iterator<size_type>(chars_start),
+                  thrust::make_counting_iterator<size_type>(chars_end),
+                  d_target_positions,
+                  target_detector);
+
+  device_span<int32_t const> d_offsets_span(d_offsets, offset_count);
+  if (target_size > 1) {
+    target_count =
+      filter_false_target_positions(target_positions, d_offsets_span, target_size, stream);
+    if (target_count == 0) {
+      // nothing to replace, copy the input column
+      return std::make_unique<cudf::column>(strings.parent(), stream, mr);
+    }
+  }
+
+  // filter out any target positions that exceed the per-row match limit
+  if (maxrepl > 0 && target_count > maxrepl) {
+    target_count = filter_maxrepl_target_positions(
+      d_target_positions, target_count, d_offsets_span, maxrepl, stream);
+  }
+
+  // build the offsets column
+  auto offsets_column = make_numeric_column(
+    data_type{type_id::INT32}, offset_count, mask_state::UNALLOCATED, stream, mr);
+  auto offsets_view     = offsets_column->mutable_view();
+  auto delta_per_target = d_repl.size_bytes() - target_size;
+  device_span<size_type const> d_target_positions_span(d_target_positions, target_count);
+  auto offsets_update_fn =
+    [d_target_positions_span, delta_per_target, chars_start] __device__(int32_t offset) -> int32_t {
+    // determine the number of target positions occurring before this offset
+    size_type const* next_target_pos_ptr = thrust::lower_bound(
+      thrust::seq, d_target_positions_span.begin(), d_target_positions_span.end(), offset);
+    size_type num_prev_targets =
+      thrust::distance(d_target_positions_span.data(), next_target_pos_ptr);
+    return offset - chars_start + delta_per_target * num_prev_targets;
+  };
+  thrust::transform(rmm::exec_policy(stream),
+                    d_offsets_span.begin(),
+                    d_offsets_span.end(),
+                    offsets_view.begin<int32_t>(),
+                    offsets_update_fn);
+
+  // build the characters column
+  auto chars_column = create_chars_child_column(strings_count,
+                                                strings.null_count(),
+                                                chars_bytes + (delta_per_target * target_count),
+                                                stream,
+                                                mr);
+  auto d_out_chars  = chars_column->mutable_view().data<char>();
+  thrust::for_each_n(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(chars_start),
+    chars_bytes,
+    target_replacer_fn{
+      d_target_positions_span, d_in_chars, d_out_chars, target_size, d_repl, chars_start});
+
+  // free the target positions buffer as it is no longer needed
+  (void)target_positions.release();
```

Comment:
`release()` doesn't free the memory. It releases ownership of device memory storage.
This could be a memory leak.

Reply:
The `device_uvector::release()` function looks like this
```
device_buffer release() noexcept { return std::move(_storage); }
```
So it returns a temporary `device_buffer` which goes no where so should be freed immediately.

Reply:
Yes, I wrote a quick test program to verify that the underlying allocation is indeed freed at the `release` call.